docs/architecture/adr-0001-app-setup-and-goals.md

```diff
@@ -0,0 +1,168 @@
+to notifications.
+
+### Impact of continuous deployment
+
+There are several consequences of the choice to aim for continuous deployment:
+
+- Continuous integration checks passing on a PR _must_ mean that the branch is
+  safe to deploy. This will require more care around migrations than usual and
+  a more thorough exercise of the application in tests than we have in Raster
+  Foundry. As long as this repository is just an API, that probably means
+  generative testing of workflows modeled through API interaction both for
+  correctness and performance regressions. I believe that we already have the
+  tools and expertise on the team to do this.
+- This repository probably shouldn't follow the `git flow` pattern, since the
+  relationship between `develop` and `master` will be pretty
+  confused (`develop = master` at all times). It is currently `develop` for
+  consistency and to avoid the appearance of having made that choice final before
+  this ADR has been reviewed.
+- We'll need to determine a good story about feature flagging and probably also
+  vertically slice user-facing development. Alternatively, if vertical slicing
+  is expected to become a bottleneck, we could attempt to find a solution through
+  team process changes, where every paired frontend and backend change requires
+  pairing. The latter alternative doesn't sound terrible to me and might have
+  the knock-on benefit of diminishing hard specialization within the team. How
+  we handle that process should be an ADR some time in the future.
+
+### Impact of keeping infrastructure as light as possible
+
+One example of how this motivation has already played out is the choice of
+how to run models. We originally considered two possibilities for how to submit
+a model run.
+
+The first option was that we demand that consumers deploy some server somewhere
+that accepts `POST`s of new imagery, responds with information about where the
+output will be written, and eventually notifies the Granary API that it can go
+fetch the imagery. The second option is what's written above.
+
+While the option we chose requires more infrastructure -- another AWS service,
+the creation of compute environments, job definitions, and roles necessary for
+running batch jobs -- this is a case where the "as possible" part of the
+motivation does a lot of work. Given that the service is designed for data
+scientists easily to run predictions against new imagery, we had to weigh that
+increase in infrastructure complexity against the presumed needs of potential
+users. Since we believe this potential user's strike zone more likely includes
+creating a docker container for running predictions than deploying a RESTful
+model-as-a-service with its own (potentially) asynchronous task runner, we
+chose for Granary to own this particular piece of infrastructure complexity.
+
+A second example is in the choice of SNS as the notification service. An
```

Comment:
Is this suggesting that we are planning to defer all notification subscription overhead to users? e.g., there will be no workflow within the application to massage the topic subscription process for users to get notifications—they will be left to their own devices to sort our that process via the AWS docs, or similar.

Reply:
I think that helps, but also since we're not thinking of this as an as-a-service thing, and instead are expecting to deploy into client's own AWS (or other cloud provider if someone will pay us to do it) account, I think we get to make slightly stronger assumptions about the accessibility / security of SNS. I think if this were as-a-service we'd have a real nightmare on our hands trying to manage who's allowed to subscribe to what topics, but the deployment story protects us there.

Reply:
Also I think the SNS topic IDs are kind of random so we get a little bit of security through that.